Made modifications to compile the code with g++ version 9.1.0 and std…

mgavin opened the pull request:
…=c++11.

include/csv_io.hpp --
Line 40:
     Check if the string_view header is available by the feature testing macro
     __cpp_lib_experimental_string_view, and include it.
     Otherwise, type-alias std::string_view to be const std::string&
     http://www.open-std.org/JTC1/SC22/WG21/docs/papers/2014/n3921.html#modifications

Line 148:
     Check if string_view is available by the feature testing macro, and
     keep the old code.
     Otherwise, initialize a const std::string& variable to the data.

Line 207:
     With -std=c++11, std::string_view is const std::string&, and isn't the
     name of a class-type anymore. Therefore, check for string_view's inclusion
     by the feature testing macro, otherwise, compare to std::string::npos

Line 261:
     Initalizers in switch-statements are a C++17 feature.
     Simply moved the initializer out. It doesn't make the code less easy to
     read, etc.
     http://open-std.org/JTC1/SC22/WG21/docs/papers/2016/p0305r1.html

mgavin committed:
Made modifications to compile the code with g++ version 9.1.0 and std=c++11.

include/csv_io.hpp --
Line 40:
     Check if the string_view header is available by the feature testing macro
     __cpp_lib_experimental_string_view, and include it.
     Otherwise, type-alias std::string_view to be const std::string&
     http://www.open-std.org/JTC1/SC22/WG21/docs/papers/2014/n3921.html#modifications

Line 148:
     Check if string_view is available by the feature testing macro, and
     keep the old code.
     Otherwise, initialize a const std::string& variable to the data.

Line 207:
     With -std=c++11, std::string_view is const std::string&, and isn't the
     name of a class-type anymore. Therefore, check for string_view's inclusion
     by the feature testing macro, otherwise, compare to std::string::npos

Line 261:
     Initalizers in switch-statements are a C++17 feature.
     Simply moved the initializer out. It doesn't make the code less easy to
     read, etc.
     http://open-std.org/JTC1/SC22/WG21/docs/papers/2016/p0305r1.html

mguid65 commented:
Change your feature test macro to 

    #if defined(__cpp_lib_string_view) || defined(__cpp_lib_experimental_string_view)
    #include <string_view>
    #endif

Don't inject names into the std namespace, this would be a disaster.

Add a macro to every line that uses `string_view` to have a version that uses `const string&` or add using directives for every container that is used and remove std:: for all std types like
    
    using std::vector;
    using std::list;
    using std::string;
    ...
    #if defined(__cpp_lib_string_view) || defined(__cpp_lib_experimental_string_view)
    #include <string_view>
    using std::string_view;
    #else
    using string_view = const string&;
    #endif

mgavin committed:
Made modifications to compile the code with g++ version 9.1.0 and std=c++11.

include/csv_io.hpp --
Line 42~:
     Rearranged includes.
     Used exclusive using directives.
     Using directives only reference types, not functions.

Line 101:
     Typo(discoverd)

Line 212:
     Took out the irrelevant check because second's type relies on tmp
     which is a std::string.

mguid65 commented:
I will accept your PR but first I will roast you for this one change 

    +     string_view dat = data;
    + #else
    +     // keep const string& (which is "string_view" now) construction, and if statement.
    +     string_view dat = (data[0] == '\"') ? data.substr(1, data.size() - 2) : data;
    + #endif
    
    -   for (const char& c : data) {
    +   for (const char& c : dat) { 

why dat, and not sv_data
